README
NAME
    Redmine::Auth - Авторизуй своих юзеров через
    редмайн! This is a private ad-hoc module, tested againts Redmine
    1.4

VERSION
    Version 0.01

SYNOPSIS
    Модуль нужен для авторизации
    пользователей на основе данных из
    редмайна. Используются данные о
    логинах\паролях и о членстве в
    проектах. Модуль тесно работает с
    Redmine::KPI, используя такие элементы как
    Element::User или как Query::Project. Поддерживает
    точно такую же инициализацию как и
    Redmine::KPI, то есть если ему скормить
    настройки доступа к Redmine, то дальше в
    коде можо плясать от него так же, как и
    от Redmine::KPI. Можно и не скармливать, тогда
    он будет просто работать, но при любой
    попытке получить данные, которые
    получает User или Query::Project, он упадет. TODO -
    отвязать от Redmine::KPI если кому-то
    понадобится

        use Redmine::Auth;

        my $auth    = new Redmine::Auth;
        $auth->dsn('DBI:mysql:dbname=redmine_test;host=localhost');
        $auth->dbUser('user');
        $auth->dbPass('pass');

        $auth->login('fedor');
        $auth->pass('123');

        say "fedor:123 works!' if $auth->auth;

        say "fedor's name is " $auth->user->param('name');

        say "fedor is manager of " . $auth->managerOf->count . 'projects';
        say "end especialy of project with id 1" if $auth->managerOf(1);

EXPORT
    none

SUBROUTINES/METHODS
  init
    Method needed for Badger::Base. Pod::Coverage - не ругайся
    пожалуйста!

  auth
    Check user credentials. Returns 0 if they are bad.

  managerOf
            say 'User is manager of project with id 1' if($auth->managerOf(1));
            say 'User is manager of ' . $auth->managerOf->count . ' projects';

    Returns bool if a numeric project id is given.

    Otherwise, returns list of projects, in wich users role is 'manager'.

    Setting Redmine::KPI fetch parameters like 'authKey' and 'url' during
    class config will give one the ability to search by project names and
    other params. Lurk for Redmine::KPI::Query::Projects documentation.

AUTHOR
    Fedor A Borshev, "<fedor at shogo.ru>"

BUGS
    Please report any bugs or feature requests to "<fedor at shogo.ru>"

SUPPORT
    You can find documentation for this module with the perldoc command.

        perldoc Redmine::Auth

LICENSE AND COPYRIGHT
    Copyright 2013 Fedor A Borshev, shogo.ru



Changes
Revision history for Redmine-Auth

0.01    Date/time
        First version, released on an unsuspecting world.



t/lib/Db.pm
package Redmine::Auth::Test::Db;
use base Redmine::Auth::Test::AABase;
use Test::More;
use strict;
use warnings;
use Redmine::Auth;

sub init_db_test : Test(3)
{
	my $self = shift;

	my $a = new Redmine::Auth;

	$self->initDsn($a);

	is($a->dsn, $self->{dsn}, 'Check setting DSN');
	is($a->dbUser, $self->{user}, 'Check setting user');
	is($a->dbPass, $self->{pass}, 'Check setting pass');
}
1;


t/lib/Main.pm
package Redmine::Auth::Test::Main;
use base Redmine::Auth::Test::AABase;
use Test::More;
use strict;
use warnings;
use Redmine::Auth;


sub auth : Test(2)
{
	my $self = shift;

	my $a = new Redmine::Auth;

	$self->initDsn($a);

	$a->login('fedor');
	$a->pass('badpass');
	is($a->auth, 0, 'Authnenticating with bad data');

	$a->pass('123');

	is($a->auth, 1, 'Authenticating with good data');
}
sub manager : Test(3)
{
	my $self = shift;

	my $a = new Redmine::Auth;

	$self->authenticate($a);

	is($a->managerOf->count, 14, 'Check getting project list');
	is($a->managerOf(1), 1, 'Check if user is manager of one project');
	is($a->managerOf(100500), 0, 'Check if user is not manager of project');
}
sub admin : Test(2)
{
	my $self = shift;
	
	my $a = new Redmine::Auth;

	$self->initDsn($a);

	$a->login('Alexey');
	$a->pass('123');
	$a->auth();

	is($a->managerOf->count, 107, 'Check getting full project list for admin user');
	is($a->managerOf(100501), 1, 'Admin is a manager if project even does not exist');
}

	
1;


t/lib/AABase.pm
package Redmine::Auth::Test::AABase;
use base Test::Class;
use strict;
use warnings;
use Test::More;
use File::Slurp qw /read_file/;

sub init : Test(setup)
{
	my $self = shift;

	my($dsn, $user, $pass);
	
	eval
	{
		$dsn	= read_file ('./t/db_data/dsn');
		$user	= read_file ('./t/db_data/user');
		$pass	= read_file ('./t/db_data/pass');

	};
	$self->SKIP_ALL('This tests need real db data in t/db_data/[dsn, user, pass]. And a redmine DB copy of course') if not $dsn or not $user or not $pass;

	chomp $dsn;
	chomp $user;
	chomp $pass;

	$self->{dsn}	= $dsn;
	$self->{user}	= $user;
	$self->{pass}	= $pass;

}
sub initDsn
{
        my $self = shift;
	my $a = shift;
	
	$self->BAIL_OUT('Bad initDsn param - need instance of Redmine::Auth') if not $a->isa('Redmine::Auth');
	$a->dsn         ($self->{dsn});
	$a->dbUser      ($self->{user});
	$a->dbPass      ($self->{pass});

	$a;
}

sub authenticate
{
	my $self = shift;
	my $a = shift;
	$self->BAIL_OUT('Bad initDsn param - need instance of Redmine::Auth') if not $a->isa('Redmine::Auth');
	$self->initDsn($a);
	$a->login('fedor');
	$a->pass('123');
	$a->auth;
}

1;


lib/Redmine/Auth.pm
package Redmine::Auth;
use Badger::Class
	base		=> 'Badger::Base',
	mutators	=> 'dsn dbUser dbPass login pass authIsOk user',
	config    	=> '',
;

use DBI;
use Digest::SHA qw/sha1_hex/;
use Redmine::KPI::Config;
use Redmine::KPI::Element::User;
use Redmine::KPI::Query::Projects;

use warnings;
use strict;

our $MANAGER_ROLE = 3;

=head1 NAME

Redmine::Auth - Авторизуй своих юзеров через редмайн! This is a private ad-hoc module, tested againts Redmine 1.4

=head1 VERSION

Version 0.01

=cut

our $VERSION = '0.01';


=head1 SYNOPSIS

Модуль нужен для авторизации пользователей на основе данных из редмайна. Используются данные о логинах\паролях и о членстве в проектах. Модуль тесно работает с Redmine::KPI, используя такие элементы как Element::User или как Query::Project. Поддерживает точно такую же инициализацию как и Redmine::KPI, то есть если ему скормить настройки доступа к Redmine, то дальше в коде можо плясать от него так же, как и от Redmine::KPI. Можно и не скармливать, тогда он будет просто работать, но при любой попытке получить данные, которые получает User или Query::Project, он упадет.
TODO - отвязать от Redmine::KPI если кому-то понадобится


    use Redmine::Auth;

    my $auth	= new Redmine::Auth;
    $auth->dsn('DBI:mysql:dbname=redmine_test;host=localhost');
    $auth->dbUser('user');
    $auth->dbPass('pass');

    $auth->login('fedor');
    $auth->pass('123');

    say "fedor:123 works!' if $auth->auth;

    say "fedor's name is " $auth->user->param('name');

    say "fedor is manager of " . $auth->managerOf->count . 'projects';
    say "end especialy of project with id 1" if $auth->managerOf(1);


=head1 EXPORT

none

=head1 SUBROUTINES/METHODS


=head2 init

Method needed for Badger::Base. Pod::Coverage - не ругайся пожалуйста!

=cut

sub init
{
	(my $self, my $config) = @_;

	$self->{config} = $self->configure($config);
	
	$self->authIsOk(0);

	$self;
}

=head2 auth

Check user credentials. Returns 0 if they are bad.

=cut

sub auth
{
	my $self = shift;

	return 1 if $self->authIsOk;
	
	my $res = $self->_dbh->prepare('SELECT id, login, admin, firstname, lastname, mail, hashed_password, salt FROM users WHERE login = ?');

	$res->execute($self->login);

	return 0 if not $res->rows;

	my $userData = $res->fetchrow_hashref;
	
	if ($userData->{hashed_password} eq sha1_hex($userData->{salt} . sha1_hex($self->pass)))
	{
		$self->authIsOk(1);
		$self->user(Redmine::KPI::Element::User->new(
				id		=> $userData->{id},
				name		=> $userData->{firstname} . ' ' . $userData->{lastname},
				firstName	=> $userData->{firstname},
				lastName	=> $userData->{lastname},
				login		=> $userData->{login},
				mail		=> $userData->{mail},
				admin		=> $userData->{admin},
				dryRun		=> 1,
				passConfigParams($self->{config}), 
			)
		);
		return 1;
	}
	return 0;
}

=head2 managerOf

	say 'User is manager of project with id 1' if($auth->managerOf(1));
	say 'User is manager of ' . $auth->managerOf->count . ' projects';
	
Returns bool if a numeric project id is given. 

Otherwise, returns list of projects, in wich users role is 'manager'.

Setting Redmine::KPI fetch parameters like 'authKey' and 'url' during class config will give one the ability to search by project names and other params. Lurk for Redmine::KPI::Query::Projects documentation.

=cut


sub managerOf
{
	my $self = shift;
	
	return $self->_projectListAsManager unless @_;

	return $self->_isManagerOf(@_);
}

sub _isManagerOf
{
	my $self = shift;
	
	return 1 if $self->user->param('admin');
	return 1 if $self->_projectListAsManager->find(shift);
	0;
}


sub _projectListAsManager { shift->_projectList('manager', $MANAGER_ROLE) }


sub _projectList
{
	my $self 	= shift;
	my $listName 	= shift;
	my $roleId	= shift;

	return if not $self->authIsOk;

	return $self->{prjLists}{$listName} if exists $self->{lists}{$listName};

	my @prjIds;
	my $res;
	if(not $self->user->param('admin'))
	{
		$res = $self->_dbh->prepare('SELECT DISTINCT (members.project_id) AS prj_id
			FROM `members` , `member_roles`
			WHERE members.user_id = ?
			AND member_roles.role_id = ?
			AND member_roles.member_id = members.id;
		');
		$res->execute($self->user->param('id'), $roleId);
	}
	else
	{
		$res = $self->_dbh->prepare('SELECT DISTINCT(id) AS prj_id FROM projects');
		$res->execute;
	}

	while (my $row = $res->fetchrow_hashref)
	{
		push @prjIds, $row->{prj_id};
	}

	@prjIds = sort { $a <=> $b } @prjIds;

	$self->{prjLists}{$listName} = Redmine::KPI::Query::Projects->new(
		project	=> \@prjIds,
		passConfigParams($self->{config}), 
	);
}
sub _dbh
{
	
	my $self = shift;

	return $self->{dbh} if exists $self->{dbh};

	$self->{dbh} = DBI->connect($self->dsn, $self->dbUser, $self->dbPass);

}


=head1 AUTHOR

Fedor A Borshev, C<< <fedor at shogo.ru> >>

=head1 BUGS

Please report any bugs or feature requests to C<< <fedor at shogo.ru> >>

=head1 SUPPORT

You can find documentation for this module with the perldoc command.

    perldoc Redmine::Auth



=head1 LICENSE AND COPYRIGHT

Copyright 2013 Fedor A Borshev, shogo.ru


=cut

1; # End of Redmine::Auth
